report_dynamic/models/report_dynamic.py

```diff
@@ -0,0 +1,271 @@
+            this.group_by_record_name = ""
+            if this.is_template:
+                continue
+            if hasattr(this.resource_ref, "name"):
+                this.group_by_record_name = this.resource_ref.name
+                continue
+            # TODO: this is plainly wrong, fix
+            this.group_by_record_name = _("{} - {}").format(
+                this.resource_ref._name, this.resource_ref.id
+            )
+
+    def get_template_xml_id(self):
+        self.ensure_one()
+        if not self.wrapper_report_id:
+            # return a default
+            return "web.external_layout"
+        record = self.env["ir.model.data"].search(
+            [("model", "=", "ir.ui.view"), ("res_id", "=", self.wrapper_report_id.id)],
+            limit=1,
+        )
+        return "{}.{}".format(record.module, record.name)
+
+    section_ids = fields.One2many("report.dynamic.section", "report_id", copy=True)
+    section_count = fields.Integer(string="Sections", compute="_compute_section_count")
+
+    @api.depends("section_ids")
+    def _compute_section_count(self):
+        for this in self:
+            this.section_count = len(this.section_ids)
+
+    def action_view_section(self):
+        self.ensure_one()
+        return {
+            "name": _("Sections"),
+            "type": "ir.actions.act_window",
+            "res_model": "report.dynamic.section",
+            "view_mode": "tree,form",
+            "target": "current",
+            "context": {"default_report_id": self.id},
+            "domain": [("id", "in", self.section_ids.ids)],
+        }
+
+    def action_wizard_lock_report(self):
+        self.ensure_one()
+        return {
+            "type": "ir.actions.act_window",
+            "res_model": "wizard.lock.report",
+            "view_mode": "form",
+            "target": "new",
+        }
+
+    def action_preview_content(self):
+        self.ensure_one()
+        action = self.env.ref("report_dynamic.report_dynamic_document_preview").read(
+            []
+        )[0]
+        return action
+
+    def action_toggle_active(self):
+        self.ensure_one()
+        self.active = not self.active
+
+    # Override create() and write() to keep
+    # resource_ref always the same with template
+    # even if template.resource_ref=False
+    @api.model
+    def create(self, values):
+        records = super().create(values)
+        for this in records:
+            if this.template_id.resource_ref and not this.res_id:
+                this.resource_ref = this.template_id.resource_ref
+                # Give a default to wrapper_report_id when
+                # user sets template_id
+                this.wrapper_report_id = this._set_wrapper_report_id(this.template_id)
+        return records
+
+    def write(self, values):
+        # If the template is changed back to a non-template
+        # (eg is_template is set to False),
+        # and the template already has children, then disallow.
+        if all(
```

Comment:
Can't you use `api.constrains`?
Also, would be nice to see these checks into separated methods.